packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts

```diff
@@ -0,0 +1,110 @@
+import { AnyJson, BareOpts } from '@polkadot/types/types/codec';
+
+export default class CENNZnetExtrinsic extends GenericExtrinsic{
+  private signaturePayloadOptions: SignatureOptions | ExtrinsicPayloadValue;
+
+  constructor(registry: Registry, value?: GenericExtrinsic | ExtrinsicValueV4 | AnyU8a | Call) {
+    super(registry, value);
+    this.signaturePayloadOptions = {
+      blockHash: undefined,
+      era: undefined,
+      genesisHash: undefined,
+      method: undefined,
+      specVersion: undefined,
+      nonce: undefined,
+      runtimeVersion: undefined,
+      signedExtensions: undefined,
+      signer: undefined,
+      tip: undefined,
+      transactionPayment: undefined,
+      transactionVersion: undefined
+    };
+  }
+
+  setNonce(value: AnyNumber){
+    this.signaturePayloadOptions.nonce = value;
+    return this;
+  }
+
+  addSignature(signer: Address | Uint8Array | string, signature: Uint8Array | string, payload: ExtrinsicPayloadValue | Uint8Array | string): GenericExtrinsic {
+    const mergeDefinedObjects = (A, B) => {
+      const res = {};
+      Object.keys({...A,...B}).map(key => {
+        res[key] = A[key] || B[key];
+      });
+      return res as ExtrinsicPayloadValue;
+    };
+
+    const mergedPayload = mergeDefinedObjects(payload, this.signaturePayloadOptions);
+    return super.addSignature(signer, signature, mergedPayload);
+  }
+
+  /**
+   * @description Sign the extrinsic with a specific keypair
+   */
+  sign(account: IKeyringPair, options: SignatureOptions): GenericExtrinsic {
+    const mergeDefinedObjects = (A, B) => {
+      const res = {};
+      Object.keys({...A,...B}).map(key => {
+          res[key] = A[key] || B[key];
+      });
+      return res as SignatureOptions;
+    };
+    const mergedSignatureOpts = mergeDefinedObjects(options, this.signaturePayloadOptions);
+    return super.sign(account, mergedSignatureOpts);
+  }
+
+  /**
+   * @description Adds payment options such as slippage and tip to the Extrinsic signature
+   */
+  paymentOpts(api: Api, feeAssetId: number | string, slippage = 0, tip = 0): this{
+    const _maxPayment = Infinity.toString();
+    const estimateFeeParams: EstimateFeeParams = {
+      encodedLength: 0, hash: undefined, isEmpty: false, registry: undefined, eq(other?: unknown): boolean {
+        return false;
+      }, toHex(isLe?: boolean): string {
+        return '';
+      }, toHuman(isExtended?: boolean): AnyJson {
+        return undefined;
+      }, toJSON(): AnyJson {
+        return undefined;
+      }, toRawType(): string {
+        return '';
+      }, toString(): string {
+        return '';
+      }, toU8a(isBare?: BareOpts): Uint8Array {
+        return undefined;
+      },
+        extrinsic : this,
+        userFeeAssetId: feeAssetId,
+        maxPayment: _maxPayment
+    }
+    const getEstimatedFee = async () => {
+      const estimatedFee = await api.derive.fees.estimateFee(estimateFeeParams)
```

Comment:
so if I understand correctly `EstimateFeeParams` doesn't impl Codec so we implemented it here? I think we can relax that requirement in the types and already did so on the `master` branch cc @KarishmaBothara 
should make this more readable

Reply:
yeah definitely so `EstimateFeeParams` shouldn't need to extend `Codec` then? looks like it was added for Prerelease/2.0.0 alpha.2 (#299)

Reply:
yeah for `derive` queries arguments don't need to have `Codec` so we decided to relax that typing

Reply:
okay sweet i'll change that